pravega-client-examples/src/main/java/io/pravega/example/streamprocessing/ExactlyOnceMultithreadedProcessor.java

```diff
@@ -0,0 +1,495 @@
+            startFromCheckpoint = startFromCheckpointName != null;
+        }
+
+        public void run() {
+            Thread.currentThread().setName("worker-" + workerIndex);
+            log.info("BEGIN");
+
+            try {
+                // Load state from checkpoint.
+                if (startFromCheckpoint) {
+                    Path checkpointDirPath = checkpointRootPath.resolve(startFromCheckpointName);
+                    Path statePath = checkpointDirPath.resolve(STATE_FILE_NAME_PREFIX + this.workerIndex);
+                    log.info("statePath={}", statePath.toString());
+                    try (FileInputStream fis = new FileInputStream(statePath.toString());
+                         ObjectInputStream ois = new ObjectInputStream(fis)) {
+                        state = (State) ois.readObject();
+                    }
+                    log.info("Loaded state {} from {}", state, statePath);
+                } else {
+                    log.info("Initializing with new state");
+                    state = new State();
+                }
+
+                try (ClientFactory clientFactory = ClientFactory.withScope(scope, controllerURI);
+                     EventStreamReader<String> reader = clientFactory.createReader(
+                             readerId,
+                             readerGroupName,
+                             new UTF8StringSerializer(),
+                             ReaderConfig.builder().build());
+                     EventStreamWriter<String> writer = clientFactory.createEventWriter(
+                             outputStreamName,
+                             new UTF8StringSerializer(),
+                             EventWriterConfig.builder()
+                                     .transactionTimeoutTime(Parameters.getTransactionTimeoutMs())
+                                     .build())) {
+
+                    Transaction<String> transaction = null;
+                    long eventCounter = 0;
+
+                    for (; ; ) {
+                        // Read input event.
+                        EventRead<String> eventRead = reader.readNextEvent(READER_TIMEOUT_MS);
+                        log.debug("readEvents: eventRead={}", eventRead);
+
+                        if (eventRead.isCheckpoint()) {
+                            // Note that next call readNextEvent will indicate to Pravega that we are done with the checkpoint.
+                            String checkpointName = eventRead.getCheckpointName();
+                            log.info("Got checkpoint {}", eventRead.getCheckpointName());
+                            Path checkpointDirPath = checkpointRootPath.resolve(checkpointName);
+                            Path transactionIdFilePath = checkpointDirPath.resolve(CHECKPOINT_TRANSACTION_ID_FILE_NAME_PREFIX + this.workerIndex);
+
+                            // Must ensure that txnId is persisted to latest state before committing transaction!
```

Comment:
Yes I believe I could do that but it is harder to understand why this would be safe to do. As it is now, only transactions that have been flushed and are ready to commit have their transaction IDs written to the files.